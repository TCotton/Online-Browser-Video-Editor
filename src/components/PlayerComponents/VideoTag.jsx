import React, {useEffect, useMemo, useState} from 'react';
import {useVideo} from 'react-use';
import {useDispatch, useSelector} from "react-redux";
import {window} from "browser-monads";
import WaveformData from "waveform-data"
import PropTypes from 'prop-types';
//import canAutoplay from 'can-autoplay';
import {selectBackward, selectForward, selectMute, selectPlay} from '../slices/playerSlice';
import {durationFn, elFn, timeFn} from '../slices/videoSlice';
import {peakFrequencyFnLeft, peakFrequencyFnRight} from '../slices/audioSlice';
import {generateThumbnail} from '../helperFunctions/generateThumbnail'
import {imageFn} from '../slices/imageSlice';
import {waveformFn} from "../slices/waveformSlice";
import videoBackground from "../../../static/video-background.png";

const VideoTag = (props) => {
    const {sources, files} = props;
    const dispatch = useDispatch();
    const [context, setContext] = useState();

    const [video, state, controls, ref] = useVideo(
        <video src={sources[0].src} id="video" poster={videoBackground}/>
    );

    useEffect(() => {
        let newRef;

        if (ref) {
            newRef = ref.current.cloneNode(true);
            const isSameNode = newRef.isSameNode(ref.current);
            if (isSameNode) return;

            newRef.muted = true;
            newRef.autoplay = true;

            let i = 0;
            const result = [];
            newRef.addEventListener('loadeddata',  () => {
                newRef.currentTime = i;
            });

            newRef.addEventListener('seeked',  () => {

                // now video has seeked and current frames will show
                // at the time as we expect
                const r = generateThumbnail(i, newRef);
                result.push(r);

                // when frame is captured, increase here by 5 seconds
                i += (newRef.duration / 5);

                // if we are not past end, seek to next interval
                if (i <= newRef.duration) {
                    // this will trigger another seeked event
                    newRef.currentTime = i;
                } else {
                    // Done!, next action
                    dispatch(imageFn(result));
                }
            });
        }
        //TODO: clean up eventListeners
    }, [files]);

    useEffect(() => {

        if (files[0]) {
            const file = files[0];
            const reader = new FileReader();
            reader.readAsArrayBuffer(file);

            reader.onload = () => {
                let audioCtx;
                if (window.webkitAudioContext) {
                    audioCtx = new window.webkitAudioContext();
                } else {
                    audioCtx = new window.AudioContext();
                }

                audioCtx.decodeAudioData(reader.result)

                    .then((audioBuffer) => {
                        const options = {
                            audio_context: audioCtx,
                            audio_buffer: audioBuffer,
                            scale: 128
                        };

                        return new Promise((resolve, reject) => {

                            WaveformData.createFromAudio(options, (err, waveform) => {
                                if (err) {
                                    reject(err);
                                } else {
                                    resolve(waveform);
                                }
                            });
                        });
                    })
                    .then(waveform => {

                        console.info(`Waveform has ${waveform.channels} channels`);
                        console.info(`Waveform has length ${waveform.length} points`);

                        const channel = waveform.channel(0);

                        const data = {
                            min_array: channel.min_array(),
                            max_array: channel.max_array(),
                            length: waveform.length,
                        }

                        dispatch(waveformFn(data));
                    }).catch((error) => {

                    throw new Error(error);
                });
            };

            reader.onerror = () => {
                throw new Error(reader.error.toString());
            };
        }
    }, [files]);

    useEffect(() => {
        //TODO: refactor into custom hook
        //TODO: move requestAnimationFrame into only component for performance reasons
        let current;
        const ctx = new (window.AudioContext || window.webkitAudioContext)();
        setContext(ctx);
        const analyser = ctx.createAnalyser();
        const source = ctx.createMediaElementSource(ref.current);
        source.connect(analyser);
        analyser.connect(ctx.destination);

        const oscillator = ctx.createOscillator();
        const channels = oscillator.channelCount;

        const splitter = ctx.createChannelSplitter(channels);
        const lAnalyser = ctx.createAnalyser();
        const rAnalyser = ctx.createAnalyser();

        source.connect(splitter);

        splitter.connect(lAnalyser, 0, 0);
        splitter.connect(rAnalyser, 1, 0);

        const lArray = new Uint8Array(lAnalyser.frequencyBinCount);
        const rArray = new Uint8Array(rAnalyser.frequencyBinCount);

        lAnalyser.getByteFrequencyData(lArray);
        rAnalyser.getByteFrequencyData(rArray);

        function requestAnimationFrameFnc() {
            current = window.requestAnimationFrame(requestAnimationFrameFnc);
            lAnalyser.getByteFrequencyData(lArray);
            rAnalyser.getByteFrequencyData(rArray);
            const peakFrequencyLeft = Math.max.apply(null, lArray);
            const peakFrequencyRight = Math.max.apply(null, rArray);
            if (peakFrequencyLeft > 0) dispatch(peakFrequencyFnLeft(peakFrequencyLeft));
            if (peakFrequencyRight > 0) dispatch(peakFrequencyFnRight(peakFrequencyRight));
        }

        requestAnimationFrameFnc();

        return () => window.cancelAnimationFrame(current);
    }, [ref]);


    useEffect(() => {
        let isStopped = false;
        if (!isStopped) {
            dispatch(durationFn(state.duration))
        }
        return () => {
            isStopped = true;
        };
    }, [state.duration]);


    useEffect(() => {
        let isStopped = false;
        if (!isStopped) {
            dispatch(timeFn(state.time))
        }
        return () => {
            isStopped = true;
        };
    }, [state.time]);


    useMemo(() => {
        //TODO: what's this for??
        dispatch(elFn(true))
    }, [ref]);


    const play = useSelector(selectPlay);
    if (play) {
        if (context && context.state === 'suspended') {
            context.resume().then(() => {
                console.log('Playback resumed successfully');
            });
        }
        controls.play();
    }
    if (!play) controls.pause();

    const back = useSelector(selectBackward);
    back.then((result) => {
        if (result) controls.seek(state.time - 0.1);
    });

    const forward = useSelector(selectForward);
    forward.then((result) => {
        if (result) controls.seek(state.time + 0.1);
    });

    const mute = useSelector(selectMute);
    mute.then((result) => {
        if (result && !state.muted) controls.mute();
        if (!result && state.muted) controls.unmute();
    });

    return (
        <>
            {video}
        </>
    )
}

VideoTag.propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    files: PropTypes.array.isRequired,
    // eslint-disable-next-line react/no-unused-prop-types
    autoPlay: PropTypes.bool.isRequired,
    // eslint-disable-next-line react/no-unused-prop-types,react/require-default-props
    controls: PropTypes.bool,
    sources: PropTypes.arrayOf(PropTypes.shape({
        src: PropTypes.string,
        type: PropTypes.string
    })).isRequired
 };

export default VideoTag;