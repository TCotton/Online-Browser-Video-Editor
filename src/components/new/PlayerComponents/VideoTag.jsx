import React, {useEffect, useMemo} from 'react';
import {useVideo} from 'react-use';
import {useDispatch, useSelector} from "react-redux";
import {window} from "browser-monads";
import {selectBackward, selectForward, selectMute, selectPlay} from '../slices/playerSlice';
import {durationFn, elFn, timeFn} from '../slices/videoSlice';
import {peakFrequencyFnLeft, peakFrequencyFnRight} from '../slices/audioSlice';

const VideoTag = (props) => {
    const {sources} = props;
    const dispatch = useDispatch();

    const [video, state, controls, ref] = useVideo(
        <video src={sources[0].src} id="video"/>
    );

    useEffect(() => {
        //TODO: refactor into custom hook
        //move requestAnimationFrame into only component for performance reasons
        let current;
        const ctx = new (window.AudioContext || window.webkitAudioContext)();
        const analyser = ctx.createAnalyser();
        const source = ctx.createMediaElementSource(ref.current);
        source.connect(analyser);
        analyser.connect(ctx.destination);

        const oscillator = ctx.createOscillator();
        const channels = oscillator.channelCount;

        const splitter = ctx.createChannelSplitter(channels);
        const lAnalyser = ctx.createAnalyser();
        const rAnalyser = ctx.createAnalyser();

        source.connect(splitter);

        splitter.connect(lAnalyser, 0, 0);
        splitter.connect(rAnalyser, 1, 0);

        const lArray = new Uint8Array(lAnalyser.frequencyBinCount);
        const rArray = new Uint8Array(rAnalyser.frequencyBinCount);

        lAnalyser.getByteFrequencyData(lArray);
        rAnalyser.getByteFrequencyData(rArray);

        function requestAnimationFrameFnc() {
            current = window.requestAnimationFrame(requestAnimationFrameFnc);
            lAnalyser.getByteFrequencyData(lArray);
            rAnalyser.getByteFrequencyData(rArray);
            const peakFrequencyLeft = Math.max.apply(null, lArray);
            const peakFrequencyRight = Math.max.apply(null, rArray);
            dispatch(peakFrequencyFnLeft(peakFrequencyLeft));
            dispatch(peakFrequencyFnRight(peakFrequencyRight));
        }

        requestAnimationFrameFnc();

        return () => window.cancelAnimationFrame(current);
    }, [ref]);


    useEffect(() => {
        let isStopped = false;
        if (!isStopped) {
            dispatch(durationFn(state.duration))
        }
        return () => {
            isStopped = true;
        };
    }, [state.duration]);


    useEffect(() => {
        let isStopped = false;
        if (!isStopped) {
            dispatch(timeFn(state.time))
        }
        return () => {
            isStopped = true;
        };
    }, [state.time]);


    useMemo(() => {
        //TODO: what's this for??
        dispatch(elFn(true))
    }, [ref]);


    const play = useSelector(selectPlay);
    if (play) controls.play();
    if (!play) controls.pause();

    const back = useSelector(selectBackward);
    back.then((result) => {
        console.log(result, 'backwards');
        if (result) controls.seek(state.time - 0.1);
    });

    const forward = useSelector(selectForward);
    forward.then((result) => {
        console.log(result, 'forward');
        if (result) controls.seek(state.time + 0.1);
    });

    const mute = useSelector(selectMute);
    mute.then((result) => {
        console.log(result, 'mute');
        if (result && !state.muted) controls.mute();
        if (!result && state.muted) controls.unmute();
    });

    return (
        <>
            {video}
        </>
    )
}


export default VideoTag;