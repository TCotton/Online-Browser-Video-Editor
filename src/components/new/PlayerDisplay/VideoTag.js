import React from 'react';
import {useVideo} from 'react-use';
import {useSelector} from "react-redux";
import {selectPlay} from '../slices/playerSlice';

const VideoTag = (props) => {
    const {autoPlay, controlz, sources} = props;

    const [video, state, controls, ref] = useVideo(
        <video src={sources[0].src}/>
    );

    const play = useSelector(selectPlay);
    play.then((result) => {
        if(result) {
            controls.play();
        }
    });

    return (
        <>
            {video}
        </>
    )
}

export default VideoTag;