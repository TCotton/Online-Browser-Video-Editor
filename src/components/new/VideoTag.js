import React from 'react';

const VideoTag = (props) => {
    const {autoPlay, controls, sources} = props;
    return (
        <video autoPlay={autoPlay} controls={controls} id="video">
            {sources.map((item) => {
                return <source key={item.src} src={item.src} type={item.type}/>
            })}
        </video>
    )
}
export default VideoTag;