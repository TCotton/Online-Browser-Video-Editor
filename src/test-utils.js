import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { createStore, combineReducers } from 'redux'
import { Provider } from 'react-redux'

import counterReducer from './components/slices/counterSlice';
import filesReducer from './components/slices/filesSlice';
import videoReducer from './components/slices/videoSlice';
import playerReducer from './components/slices/playerSlice';
import audioReducer from './components/slices/audioSlice';
import imageReducer from './components/slices/imageSlice';
import waveformReducer from './components/slices/waveformSlice';

function render(
    ui,
    {
        initialState,
        store = createStore(combineReducers({
            counter: counterReducer,
            files:  filesReducer,
            video: videoReducer,
            player: playerReducer,
            audio: audioReducer,
            images: imageReducer,
            waveform: waveformReducer,
        }), initialState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {
        return <Provider store={store}>{children}</Provider>
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
// eslint-disable-next-line import/no-extraneous-dependencies
export * from '@testing-library/react'
// override render method
export { render }