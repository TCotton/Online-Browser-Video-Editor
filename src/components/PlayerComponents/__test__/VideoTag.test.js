import React from "react";
import VideoTag from "../VideoTag";
import {render} from "../../../test-utils";

describe('VideoTag', () => {

    const mockConnect = jest.fn();
    const mockcreateMediaElementSource = jest.fn(() => {
        return {
            connect: mockConnect
        }
    });
    const mockgetByteFrequencyData = jest.fn();
    const mockcreateAnalyser = jest.fn(() => {
        return {
            connect: mockConnect,
            frequencyBinCount: [0, 1, 2],
            getByteFrequencyData: mockgetByteFrequencyData,
        }
    });
    const mockcreateOscillator = jest.fn(() => {
        return {
            channelCount: 2
        }
    });
    const mockChannelSplitterConnect = jest.fn(n => n);
    const mockcreateChannelSplitter = jest.fn(() => {
        return {
            connect: mockChannelSplitterConnect
        }
    });
    const mockaudioContext = jest.fn(() => {
        return {
            createAnalyser: mockcreateAnalyser,
            createMediaElementSource: mockcreateMediaElementSource,
            createOscillator: mockcreateOscillator,
            createChannelSplitter: mockcreateChannelSplitter,
        }
    });

    const video = new File(['(⌐□_□)'], 'video.mp4', {type: 'video/mp4'});
    //const blob = new Blob([range(1024)], {type: "video/mp4"});

    const props = {
        files: [video],
        autoPlay: true,
        controls: true,
        sources: [{
            src: 'blob:http://localhost:8000/c22f16b2-09e1-4fd1-a2e1-3402d215b992',
            type: 'video/mp4',
        }]
    }

    const play = jest.fn();
    const pause = jest.fn();
    const load = jest.fn();
    const seek = jest.fn();

    beforeEach(() => {
        window.AudioContext = mockaudioContext;
        window.HTMLMediaElement.prototype.load = load;
        window.HTMLMediaElement.prototype.play = play;
        window.HTMLMediaElement.prototype.pause = pause;
        window.HTMLMediaElement.prototype.seek = seek;
    });

    it('should render successfully', () => {
        const {container} = render(<VideoTag {...props} />);
        expect(container.querySelector("[data-testid=\"video\"]")).not.toBeNull();
    });

    it('should match snapshot', () => {
        const {baseElement} = render(<VideoTag {...props} />);
        expect(baseElement).toMatchSnapshot();
    });

    it('should call audioContext API', () => {
        render(<VideoTag {...props} />);
        expect(mockcreateAnalyser).toBeCalled();
        expect(mockcreateMediaElementSource).toBeCalled();
        expect(mockConnect).toBeCalled();
        expect(mockcreateOscillator).toBeCalled();
        expect(mockcreateChannelSplitter).toBeCalled();
        expect(mockChannelSplitterConnect).toBeCalled();
        expect(mockgetByteFrequencyData).toBeCalled();
    });

    it('should call window.HTMLMediaElement play on initial render', () => {
        render(<VideoTag {...props} />);
        expect(play).toBeCalled();
    });

    it('should call window.HTMLMediaElement pause on if state is changed', () => {
        render(<VideoTag {...props}/>, {
            initialState: {
                player: {
                    play: false,
                    stop: false,
                    forward: false,
                    backward: false,
                    mute: false,
                }
            }
        });
        expect(pause).toBeCalled();
    });
});
