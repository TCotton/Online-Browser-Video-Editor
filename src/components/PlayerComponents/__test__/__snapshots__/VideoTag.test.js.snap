// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoTag should match snapshot 1`] = `
<body>
  <div>
    <video
      data-testid="video"
      id="video"
      poster="test-file-stub"
      src="blob:http://localhost:8000/c22f16b2-09e1-4fd1-a2e1-3402d215b992"
    />
  </div>
</body>
`;
